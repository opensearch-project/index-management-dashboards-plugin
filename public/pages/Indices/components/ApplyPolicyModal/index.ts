/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import ApplyPolicyModal from "./ApplyPolicyModal";

export default ApplyPolicyModal;
