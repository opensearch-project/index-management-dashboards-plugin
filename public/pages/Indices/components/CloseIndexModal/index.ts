/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import CloseIndexModal from "./CloseIndexModal";

export default CloseIndexModal;
