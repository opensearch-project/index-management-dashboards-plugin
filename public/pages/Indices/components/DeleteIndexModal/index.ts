/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import DeleteIndexModal from "./DeleteIndexModal";

export default DeleteIndexModal;
