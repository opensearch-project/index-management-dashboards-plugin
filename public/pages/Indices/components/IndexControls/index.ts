/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import IndexControls from "./IndexControls";

export default IndexControls;
