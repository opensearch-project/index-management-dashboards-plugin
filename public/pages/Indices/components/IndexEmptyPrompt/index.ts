/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import IndexEmptyPrompt from "./IndexEmptyPrompt";

export default IndexEmptyPrompt;
