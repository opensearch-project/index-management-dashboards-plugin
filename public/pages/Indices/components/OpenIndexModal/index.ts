/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import OpenIndexModal from "./OpenIndexModal";

export default OpenIndexModal;
