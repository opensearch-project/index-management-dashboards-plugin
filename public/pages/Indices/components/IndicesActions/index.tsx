/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */
import React, { useCallback, useContext, useMemo, useState } from "react";
import {
  // @ts-ignore
  Criteria,
  // @ts-ignore
  Pagination,
  EuiButton,
  EuiContextMenu,
} from "@elastic/eui";

import { ManagedCatIndex } from "../../../../../server/models/interfaces";
import ApplyPolicyModal from "../../components/ApplyPolicyModal";
import SimplePopover from "../../../../components/SimplePopover";
import { ModalConsumer } from "../../../../components/Modal";
import { CoreServicesContext } from "../../../../components/core_services";
import DeleteIndexModal from "../../components/DeleteIndexModal";
import ReindexFlyout from "../ReindexFlyout";
import { REQUEST } from "../../../../../utils/constants";
import { ReindexRequest } from "../../models/interfaces";
import IndexService from "../../../../services/IndexService";
import CommonService from "../../../../services/CommonService";
import { ServicesContext } from "../../../../services";

interface IndicesActionsProps {
  selectedItems: ManagedCatIndex[];
  onDelete?: () => void;
  indexService: IndexService;
  commonService: CommonService;
}

export default function IndicesActions(props: IndicesActionsProps) {
  const { selectedItems, onDelete, commonService } = props;
  const [deleteIndexModalVisible, setDeleteIndexModalVisible] = useState(false);
  const [isReindexFlyoutVisible, setIsReindexFlyoutVisible] = useState(false);
  const context = useContext(CoreServicesContext);
  const services = useContext(ServicesContext);

  const onDeleteIndexModalClose = () => {
    setDeleteIndexModalVisible(false);
  };

  const onDeleteIndexModalConfirm = useCallback(async () => {
    const result = await services?.commonService.apiCaller({
      endpoint: "indices.delete",
      data: {
        index: selectedItems.map((item) => item.index).join(","),
      },
    });
    if (result && result.ok) {
      onDeleteIndexModalClose();
      context?.notifications.toasts.addSuccess("Delete successfully");
      onDelete && onDelete();
    } else {
      context?.notifications.toasts.addDanger(result?.error || "");
    }
  }, [selectedItems, onDelete, services, context, onDeleteIndexModalClose]);

  const onReindexConfirmed = async (reindexRequest: ReindexRequest) => {
    let res = await commonService.apiCaller({
      endpoint: "reindex",
      method: REQUEST.POST,
      data: reindexRequest,
    });
    if (res.ok) {
      // @ts-ignore
      context?.notifications.toasts.addSuccess(`Reindex triggered successfully with taskId ${res.response.task}`);
      onCloseReindexFlyout();
    } else {
      context?.notifications.toasts.addDanger(`Reindex operation error happened ${res.error}`);
    }
  };

  const onCloseReindexFlyout = () => {
    setIsReindexFlyoutVisible(false);
  };

  const renderKey = useMemo(() => Date.now(), [selectedItems]);

  return (
    <>
      <ModalConsumer>
        {({ onShow }) => (
          <SimplePopover
            data-test-subj="More Action"
            panelPaddingSize="none"
            button={
              <EuiButton iconType="arrowDown" iconSide="right">
                Actions
              </EuiButton>
            }
          >
            <EuiContextMenu
              initialPanelId={0}
              // The EuiContextMenu has bug when testing in jest
              // the props change won't make it rerender
              key={renderKey}
              panels={[
                {
                  id: 0,
                  items: [
                    {
                      name: "Delete",
                      disabled: !selectedItems.length,
                      "data-test-subj": "Delete Action",
                      onClick: () => setDeleteIndexModalVisible(true),
                    },
                    {
                      name: "Apply policy",
                      disabled: !selectedItems.length,
                      "data-test-subj": "Apply policyButton",
                      onClick: () =>
                        onShow(ApplyPolicyModal, {
                          indices: selectedItems.map((item: ManagedCatIndex) => item.index),
                          core: context,
                        }),
                    },
                    {
                      name: "Reindex",
                      disabled: !selectedItems.length,
                      "data-test-subj": "Reindex Action",
                      onClick: () => setIsReindexFlyoutVisible(true),
                    },
                  ],
                },
              ]}
            />
          </SimplePopover>
        )}
      </ModalConsumer>
      <DeleteIndexModal
        selectedItems={selectedItems.map((item) => item.index)}
        visible={deleteIndexModalVisible}
        onClose={onDeleteIndexModalClose}
        onConfirm={onDeleteIndexModalConfirm}
      />

      {isReindexFlyoutVisible && (
        <ReindexFlyout
          {...props}
          onCloseFlyout={onCloseReindexFlyout}
          sourceIndices={selectedItems.map((item) => item.index)}
          onReindexConfirmed={onReindexConfirmed}
        />
      )}
    </>
  );
}
