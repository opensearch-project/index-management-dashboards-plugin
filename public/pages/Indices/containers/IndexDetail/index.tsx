/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { useContext, useEffect, useMemo, useState } from "react";
import {
  EuiButtonEmpty,
  EuiCopy,
  EuiTabbedContent,
  EuiDescriptionList,
  EuiDescriptionListTitle,
  EuiDescriptionListDescription,
  EuiFlexGrid,
  EuiFlexItem,
  EuiSpacer,
  EuiFlexGroup,
  EuiButton,
  EuiBasicTable,
} from "@elastic/eui";
import { get } from "lodash";
import { Link } from "react-router-dom";
import { IndexItem } from "../../../../../models/interfaces";
import { Modal } from "../../../../components/Modal";
import IndicesActions, { IndicesActionsProps } from "../IndicesActions";
import { ManagedCatIndex } from "../../../../../server/models/interfaces";
import { IndicesUpdateMode, ROUTES } from "../../../../utils/constants";
import JSONEditor from "../../../../components/JSONEditor";
import { ServicesContext } from "../../../../services";
import { BrowserServices } from "../../../../models/interfaces";

export interface IndexDetailModalProps extends Pick<IndicesActionsProps, "onDelete"> {
  index: string;
  record: ManagedCatIndex;
}

interface IFinalDetail extends ManagedCatIndex, IndexItem {}

const OVERVIEW_DISPLAY_INFO: {
  label: string;
  value: string | React.FunctionComponent<{ detail: IFinalDetail }>;
}[] = [
  {
    label: "Index name",
    value: "index",
  },
  {
    label: "Health",
    value: "health",
  },
  {
    label: "Status",
    value: "status",
  },
  {
    label: "Creation date",
    value: ({ detail }) => <span>{new Date(parseInt(detail.settings?.index.creation_date || "0")).toLocaleString()}</span>,
  },
  {
    label: "Total size",
    value: ({ detail }) => <span>{detail["store.size"]}</span>,
  },
  {
    label: "Primaries size",
    value: ({ detail }) => <span>{detail["pri.store.size"]}</span>,
  },
  {
    label: "Total documents",
    value: ({ detail }) => <span>{detail["docs.count"]}</span>,
  },
  {
    label: "Deleted documents",
    value: ({ detail }) => <span>{detail["docs.deleted"]}</span>,
  },
  {
    label: "Primaries",
    value: "pri",
  },
  {
    label: "Replicas",
    value: "rep",
  },
  {
    label: "Index blocks",
    value: () => <span>-</span>,
  },
  {
    label: "Managed by policy",
    value: ({ detail }) => (
      <span>
        {detail.managedPolicy ? <Link to={`${ROUTES.POLICY_DETAILS}?id=${detail.managedPolicy}`}>{detail.managedPolicy}</Link> : "-"}
      </span>
    ),
  },
];

export default function IndexDetail(props: IndexDetailModalProps) {
  const { index, record, onDelete } = props;
  const [visible, setVisible] = useState(false);
  const [detail, setDetail] = useState({} as IndexItem);
  const finalDetail: IFinalDetail = useMemo(
    () => ({
      ...record,
      ...detail,
    }),
    [record, detail]
  );
  const services = useContext(ServicesContext) as BrowserServices;

  useEffect(() => {
    if (visible) {
      services.commonService
        .apiCaller<Record<string, IndexItem>>({
          endpoint: "indices.get",
          data: {
            index,
          },
        })
        .then((res) => {
          if (!res.ok) {
            return res;
          }

          return {
            ...res,
            response: res.response[index],
          };
        })
        .then((res) => {
          if (res && res.ok) {
            setDetail(res.response);
          }
        });
    }
  }, [visible]);
  return (
    <>
      <EuiCopy textToCopy={index}>
        {(copy) => <EuiButtonEmpty size="xs" flush="right" iconType="copyClipboard" onClick={copy} color="text"></EuiButtonEmpty>}
      </EuiCopy>
      <EuiButtonEmpty onClick={() => setVisible(true)} data-test-subj={`view-index-detail-button-${index}`}>
        {index}
      </EuiButtonEmpty>
      <Modal.SimpleModal
        visible={visible}
        maxWidth={false}
        style={{
          width: "60vw",
        }}
        onClose={() => setVisible(false)}
        title={
          <div style={{ display: "flex", justifyContent: "space-between" }}>
            <span>{index}</span>
            <IndicesActions selectedItems={[record]} onDelete={onDelete} />
          </div>
        }
        content={
          <EuiTabbedContent
            display="condensed"
            tabs={[
              {
                id: "index-detail-modal-overview",
                name: "Overview",
                content: (
                  <EuiDescriptionList>
                    <EuiSpacer />
                    <EuiFlexGrid columns={3}>
                      {OVERVIEW_DISPLAY_INFO.map((item) => {
                        let valueContent = null;
                        if (typeof item.value === "string") {
                          valueContent = <span>{get(finalDetail, item.value)}</span>;
                        } else {
                          const ValueComponent = item.value;
                          valueContent = <ValueComponent detail={finalDetail} />;
                        }
                        return (
                          <EuiFlexItem key={item.label} data-test-subj={`index-detail-overview-item-${item.label}`}>
                            <EuiDescriptionListTitle>{item.label}</EuiDescriptionListTitle>
                            <EuiDescriptionListDescription>{valueContent}</EuiDescriptionListDescription>
                          </EuiFlexItem>
                        );
                      })}
                    </EuiFlexGrid>
                  </EuiDescriptionList>
                ),
              },
              {
                id: "index-detail-modal-settings",
                name: "Settings",
                content: (
                  <>
                    <EuiSpacer />
                    <EuiFlexGroup justifyContent="spaceBetween" alignItems="flexEnd">
                      <EuiFlexItem grow={false}>
                        <h6>Advanced index settings</h6>
                      </EuiFlexItem>
                      <EuiFlexItem grow={false}>
                        <Link to={`${ROUTES.CREATE_INDEX}/${index}/${IndicesUpdateMode.settings}`}>
                          <EuiButton size="s" data-test-subj="detail-modal-edit">
                            Edit
                          </EuiButton>
                        </Link>
                      </EuiFlexItem>
                    </EuiFlexGroup>
                    <EuiSpacer />
                    <JSONEditor
                      key="index-detail-modal-settings-editor"
                      readOnly
                      value={JSON.stringify(finalDetail.settings || {}, null, 2)}
                    />
                  </>
                ),
              },
              {
                id: "index-detail-modal-mappings",
                name: "Mappings",
                content: (
                  <>
                    <EuiSpacer />
                    <EuiFlexGroup justifyContent="spaceBetween" alignItems="flexEnd">
                      <EuiFlexItem grow={false}>
                        <h6>Index mappings</h6>
                      </EuiFlexItem>
                      <EuiFlexItem grow={false}>
                        <Link to={`${ROUTES.CREATE_INDEX}/${index}/${IndicesUpdateMode.mappings}`}>
                          <EuiButton size="s" data-test-subj="detail-modal-edit">
                            Edit
                          </EuiButton>
                        </Link>
                      </EuiFlexItem>
                    </EuiFlexGroup>
                    <EuiSpacer />
                    <JSONEditor
                      key="index-detail-modal-mappings-editor"
                      readOnly
                      value={JSON.stringify(finalDetail.mappings || {}, null, 2)}
                    />
                  </>
                ),
              },
              {
                id: `index-detail-modal-${IndicesUpdateMode.alias}`,
                name: "Alias",
                content: (
                  <>
                    <EuiSpacer />
                    <EuiFlexGroup justifyContent="spaceBetween" alignItems="flexEnd">
                      <EuiFlexItem grow={false}>
                        <h6>Index alias</h6>
                      </EuiFlexItem>
                      <EuiFlexItem grow={false}>
                        <Link to={`${ROUTES.CREATE_INDEX}/${index}/${IndicesUpdateMode.alias}`}>
                          <EuiButton size="s" data-test-subj="detail-modal-edit">
                            Edit
                          </EuiButton>
                        </Link>
                      </EuiFlexItem>
                    </EuiFlexGroup>
                    <EuiSpacer />
                    <EuiBasicTable
                      rowHeader="alias"
                      noItemsMessage="No alias found"
                      items={Object.keys(finalDetail.aliases || {}).map((item) => ({ alias: item }))}
                      columns={[
                        {
                          field: "alias",
                          name: "Alias name",
                          render: (val: string, record: { alias: string }) => <Link to="somewhereto">{val}</Link>,
                        },
                      ]}
                    />
                  </>
                ),
              },
            ]}
          />
        }
      />
    </>
  );
}
