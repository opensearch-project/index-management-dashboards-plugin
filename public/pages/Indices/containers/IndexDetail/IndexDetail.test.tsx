/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from "react";
import "@testing-library/jest-dom/extend-expect";
import { render, waitFor } from "@testing-library/react";
// @ts-ignore
import userEvent from "@testing-library/user-event";
import { browserServicesMock, coreServicesMock } from "../../../../../test/mocks";
import IndexDetail, { IndexDetailModalProps } from "./index";
import { ModalProvider } from "../../../../components/Modal";
import { ServicesContext } from "../../../../services";
import { CoreServicesContext } from "../../../../components/core_services";

function renderWithRouter(props: IndexDetailModalProps) {
  return {
    ...render(
      <CoreServicesContext.Provider value={coreServicesMock}>
        <ServicesContext.Provider value={browserServicesMock}>
          <ModalProvider>
            <IndexDetail {...props} />
          </ModalProvider>
        </ServicesContext.Provider>
      </CoreServicesContext.Provider>
    ),
  };
}

describe("container <IndexDetail /> spec", () => {
  it("render the component", async () => {
    browserServicesMock.commonService.apiCaller = jest.fn().mockResolvedValue({
      ok: true,
      response: {
        test_index: {
          aliases: {},
          mappings: {},
          settings: {
            index: {
              number_of_shards: "1",
              number_of_replicas: "1",
              provided_name: "test_index",
            },
          },
        },
      },
    });
    const { container, getByTestId } = renderWithRouter({
      index: "test_index",
      record: {
        "docs.count": "5",
        "docs.deleted": "2",
        health: "green",
        index: "test_index",
        pri: "1",
        "pri.store.size": "100KB",
        rep: "0",
        status: "open",
        "store.size": "100KB",
        uuid: "some_uuid",
        managed: "",
        managedPolicy: "",
        data_stream: "",
      },
      onDelete: () => null,
    });

    await waitFor(() => {
      expect(container.firstChild).toMatchSnapshot();
    });

    userEvent.click(getByTestId("view-index-detail-button-test_index"));

    await waitFor(() => {
      expect(document.querySelector("#index-detail-modal-overview")).not.toBeNull();
      expect(browserServicesMock.commonService.apiCaller).toBeCalledTimes(1);
      expect(browserServicesMock.commonService.apiCaller).toBeCalledWith({
        endpoint: "indices.get",
        data: {
          index: "test_index",
        },
      });
    });
  });
});
