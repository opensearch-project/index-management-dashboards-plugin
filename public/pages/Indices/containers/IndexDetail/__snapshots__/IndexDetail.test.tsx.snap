// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`container <IndexDetail /> spec render the component 1`] = `
<span
  class="euiToolTipAnchor"
>
  <button
    class="euiButtonEmpty euiButtonEmpty--text euiButtonEmpty--xSmall euiButtonEmpty--flushRight"
    type="button"
  >
    <span
      class="euiButtonContent euiButtonEmpty__content"
    >
      EuiIconMock
      <span
        class="euiButtonEmpty__text"
      />
    </span>
  </button>
</span>
`;
